website/www/site/content/en/blog/apache-beam-flink-and-kubernetes.md

```diff
@@ -0,0 +1,355 @@
+
+Autoscaler needs to scale up the number of task managers to drain the lag and keep up with the throughput. It will also scale down the minimum number of resources required to process the incoming traffic to save cost. We need to do this frequently while keeping the processing disruption to minimum.
+
+We have tuned the autoscaler intensively, so that we can meet the SLA for latency, and make trade off for cost. We also make the autoscaler application specific so that we can meet specific needs for certain applications. Every decision has a hidden cost.  We dive into the great details of autoscaler in Part 2 of this blog.
+
+
+# Creating a Client Library for Steaming Job Development
+
+To deploy the job via Flink Kubernetes Operator, one has to be knowledgeable enough in the workings of kubernetes. The steps to create a single Flink job is:
+
+
+
+1. Define a yaml file with proper specifications. This is an example:
+
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/flink-deployment-yaml.png"
+alt="Flink Deployment Yaml">
+
+
+2. SSH inside your Flink cluster and run the command:
+
+    ```
+kubectl create -f job1.yaml
+```
+
+
+3. Check the status of the job:
+
+    ```
+kubectl get flinkdeployment job1
+```
+
+
+
+This process impacts our scalability. We frequently update our jobs so It is not feasible to do this manually for thousands of running jobs, as it's highly error prone and slow. One wrong spacing in the yaml can fail the deployment. This approach also acts as a barrier to innovation as one must know kubernetes to interact with Flink jobs.
+
+To solve these, we decided to build a library to provide an interface for any teams and applications wanting to start, delete, update or get status of their jobs.
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/fko-library.png"
+alt="Flink Kubernetes Operator Library">
+
+
+This library extends Fabric8 client and FlinkDeployment CRD. FlinkDeployment CRD is exposed by Flink Kubernetes Operator. CRD lets you store and retrieve structured data. By extending the CRD, we get access to POJO, making it easier to manipulate the yaml file.
+
+It will:
+
+
+
+1. Authenticate - ensure you are allowed to perform actions on the Flink cluster
+2. Validate (Fetches template from AWS/GCS for validation) - will take user variable input and validate it against the policy and rules and yaml format.
+3. Execute the action - convert the java call to invoke kubernetes operation
+
+Lessons learned:
+
+
+
+1. App specific operator service - at our large scale, the operator was not able to handle such a large amount of jobs. Kubernetes calls started to time out and failed. To solve this, we created multiple operators (~4) in high traffic regions to handle each application.
```

Comment:
```suggestion
1. App specific operator service: At our large scale, the operator was unable to handle such a large number of jobs. Kubernetes calls started to time out and fail. To solve this problem, we created multiple operators (about 4) in high-traffic regions to handle each application.
```

Reply:
Did you create four operators in each high traffic region, or four total?

Reply:
We created high traffic regions more than one operator. One biggest region has 4 operator to handle jobs.